rfcs/20241008-graph-api-swish/README.md

```diff
@@ -0,0 +1,185 @@
+- OpenVINO supports Swish via the Swish operation [[#8]][8]. Unlike PyTorch's
+  SiLU operation, OpenVINO's Swish also accepts a scalar input `Beta` as the
+  multiplication `factor` for Sigmoid.
+- For ONNX, a PR is working in progress to add Swish operation [[#9]][9].
+- oneDNN supports Swish as an algorithm of eltwise primitive [[#10]][10] which
+  accepts a scalar `alpha` in primitive descriptor creation as the
+  multiplication `factor` for Sigmoid.
+- cuDNN backend API supports Swish as a mode (`CUDNN_POINTWISE_SWISH_FWD`) of
+  its Pointwise operation [[#11]][11] and accepts attribute
+  `CUDNN_ATTR_POINTWISE_SWISH_BETA` as the multiplication `factor`.
+- Please note that, even PyTorch has SiLU operation, there are still many model
+  scripts choosing to implement swish with a composition of smaller operations
+  [[#12]][12].
+
+## Proposals
+
+### Option 1: Support Swish via Sigmoid and Multiply operation
+
+As indicated by the formula of Swish, the proposal is to support it via the
+combination of Sigmoid and Multiply operations which are already supported in
+oneDNN Graph API.
+
+- [Sigmoid operation](https://oneapi-src.github.io/oneDNN/dev_guide_op_sigmoid.html)
+- [Multiply operation](https://oneapi-src.github.io/oneDNN/dev_guide_op_multiply.html)
+
+With that, a Swish operation with default `factor` can ben programed as below:
+
+```cpp
+using namespace dnnl::graph;
+
+graph swish = graph(engine::kind::cpu);
+
+logical_tensor src = logical_tensor(ID_SRC, dt, shape);
+logical_tensor res = logical_tensor(ID_RES, dt, shape);
+logical_tensor dst = logical_tensor(ID_DST, dt, shape);
+
+op sig = op(ID_SIG, op::kind::Sigmoid, "sig");
+sig.add_input(src);
+sig.add_output(res);
+
+op mul = op(ID_MUL, op::kind::Multiply, "mul");
+mul.add_inputs({src, res});
+mul.add_output(dst);
+
+swish.add_op(sig);
+swish.add_op(mul);
+swish.finalize();
+```
+
+Pros:
+
+- There is no need to define and maintain a new operation in oneDNN Graph API.
+- Composition of smaller operations makes it possible and scalable to extend whe
+  the activation has more variants or flavors in the future.
+- The approach of composition of Multiply and Sigmoid is also adopted in models
+  as mentioned above.
+
+Cons:
+
+- Compared to a dedicate Swish operation, this proposal requires more users code
+  (at least one more logical tensor and one more operation).
+- It also requires complex logic in the backend to detect `Sigmoid + Multiply`
+  and map to the existing Swish kernels in oneDNN. It requires the input of
+  Sigmoid and the second input of Multiply to be the same tensor.
+- Considering that SiLU is a built-in operation in PyTorch, mapping it to two
+  operations in oneDNN Graph is troublesome for some integrations.
+- Currently, oneDNN Graph Sigmoid operation does not support a multiplication
+  `factor`. We may need to extend either the proposed Swish graph or the Sigmoid
+  operation to support cases where `factor != 1.f`.
```

Comment:
Since PyTorch's SiLU operation only supports factor = 1.f, it can be simply decomposed to Sigmoid + Multiply. When we dispatch these pattern to oneDNN swish kernel, we just set alpha = 1.f. It works perfectly.
But if we want to support factor != 1.f for other cases, the operation will be decomposed into Multiply + Sigmoid + Multiply with the first Multiply to do `factor * x`. To dispatch this pattern into oneDNN swish kernel, we have more things to consider:
- We need to check that the factor input of the first Multiply should be a scalar.
- The value of the scalar should be const among iterations.
- The value of the scalar should be known at compilation stage as it's required to create primitive descriptors. But for now, the compile() API only accepts the logical tensors of input tensors, not the values of input tensors.

Reply:
THat is correct. For that case, you would need an extra attribute to Sigmoid with alpha value.

Reply:
Thanks! Or, can we simply rely on binary + eltwise post-op + binary post-op for the cases where we don't know factor at compilation?

Reply:
That is an option as well. THough I would expect for the specific case of switch, the alpha parameter is constant, and using switch eltwise would be faster.
The other thing is GPU support. alpha might not be a memory object on GPU.